apps/extension/src/routes/offscreen/build.ts

```diff
@@ -0,0 +1,80 @@
+import {
+  InternalMessageHandler,
+  InternalResponse,
+} from '@penumbra-zone/types/src/internal-msg/shared';
+import { ActionBuildMessage } from './types';
+import {
+  Action,
+  ActionPlan,
+  TransactionPlan,
+  WitnessData,
+} from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/transaction/v1alpha1/transaction_pb';
+
+const spawnWorker = (
+  transactionPlan: TransactionPlan,
+  actionPlan: ActionPlan,
+  witness: WitnessData,
+  fullViewingKey: string,
+  key_type: string,
+): Promise<Action> => {
+  return new Promise((resolve, reject) => {
+    const worker = new Worker(new URL('../web-worker/router.ts', import.meta.url));
+
+    // Set up event listener to recieve messages from the web worker
+    worker.addEventListener(
+      'message',
+      function (e) {
+        resolve(e.data as Action);
+      },
+      false,
+    );
+
+    // Set up error handling
+    worker.addEventListener('error', function (error) {
+      console.error('Error in worker:', error);
+      reject(error);
+    });
+
+    // Send data to web worker
+    worker.postMessage({
+      type: 'worker',
+      data: {
+        transactionPlan,
+        actionPlan,
+        witness,
+        fullViewingKey,
+        key_type,
+      },
+    });
+  });
+};
+
+export const buildActionHandler: InternalMessageHandler<ActionBuildMessage> = async (
+  jsonReq,
+  responder,
+): Promise<InternalResponse<ActionBuildMessage>> => {
+  // Destructure the data object to get individual fields
+  const { transactionPlan, actionPlan, witness, fullViewingKey, key_type } = jsonReq;
+
+  // Array to store promises for each worker
+  const workerPromises: Promise<Action>[] = [];
+
+  // Spawn web workers
+  for (let i = 0; i < actionPlan.length; i++) {
+    workerPromises.push(
+      spawnWorker(transactionPlan, actionPlan[i]!, witness, fullViewingKey, key_type[i]!),
```

Comment:
If you use a `for of...` loop, you don't need to have this array indexing thing.

Reply:
If `actionPlan[i]` is paired with `key_type[i]`, it would be nice to have a data structure that combines them both so they are always together and the array length/order doesn't matter.

Reply:
changed to `for (const [i, action] of actionPlan.entries())`